core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,593 @@
+    slot_number: Slot,
+    state: u64,
+}
+
+const MOCK_VOTE_HEADER_SIZE: usize = std::mem::size_of::<MockVotePacketHeader>();
+
+/// The actual alpenglow votor packets are all smaller than this,
+/// but this is deliberately overtuned to model the worst case.
+const MOCK_VOTE_PACKET_SIZE: usize = 512;
+
+impl MockVotePacketHeader {
+    fn from_bytes_mut(buf: &mut [u8]) -> &mut Self {
+        bytemuck::from_bytes_mut::<MockVotePacketHeader>(&mut buf[..MOCK_VOTE_HEADER_SIZE])
+    }
+    fn from_bytes(buf: &[u8]) -> &Self {
+        bytemuck::from_bytes::<MockVotePacketHeader>(&buf[..MOCK_VOTE_HEADER_SIZE])
+    }
+}
+const EXIT_CHECK_INTERVAL: Duration = Duration::from_millis(100);
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum Command {
+    SendNotarize(Slot),
+    SendNotarizeCertificate(Slot),
+    SendFinalize(Slot),
+    SendFinalizeCertificate(Slot),
+}
+
+impl MockAlpenglowConsensus {
+    pub(crate) fn new(
+        alpenglow_socket: UdpSocket,
+        cluster_info: Arc<ClusterInfo>,
+        epoch_specs: EpochSpecs,
+    ) -> Self {
+        info!("Mock Alpenglow consensus is enabled");
+        let socket = Arc::new(alpenglow_socket);
+        let (command_sender, vote_command_receiver) = bounded(4);
+        let shared_state = Arc::new(Mutex::new(SharedState {
+            should_exit: false,
+            current_slot_start: Instant::now(),
+            peers: HashMap::new(),
+            current_slot: 0,
+            total_staked: 0,
+            alpenglow_state: AgStateMachine::default(),
+        }));
+        Self {
+            state: shared_state.clone(),
+            listener_thread: thread::spawn({
+                let shared_state = shared_state.clone();
+                let socket = socket.clone();
+                let cluster_info = cluster_info.clone();
+                let epoch_specs = epoch_specs.clone();
+                let command_sender = command_sender.clone();
+                move || {
+                    Self::listener_thread(
+                        shared_state,
+                        cluster_info,
+                        epoch_specs,
+                        socket,
+                        command_sender,
+                    )
+                }
+            }),
+            sender_thread: thread::spawn({
+                let epoch_specs = epoch_specs.clone();
+                let cluster_info = cluster_info.clone();
+                move || {
+                    Self::sender_thread(
+                        shared_state,
+                        cluster_info,
+                        epoch_specs,
+                        socket.clone(),
+                        vote_command_receiver,
+                    )
+                }
+            }),
+            epoch_specs,
+            command_sender,
+            cluster_info,
+        }
+    }
+
+    fn prepare_to_receive(&mut self, slot: Slot) {
+        let staked_nodes = self.epoch_specs.current_epoch_staked_nodes();
```

Comment:
This is the staked_nodes for the epoch the root is in? It might be different from the Epoch current_slot is in.